tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/provider/BlobStoreRepository.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.mledger.offload.jcloud.provider;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.jclouds.blobstore.BlobStore;
+
+/**
+ * Centralized storage location for all the Blobstore instances
+ * created and used for tiered-storage.
+ * <p>
+ * Users should first check to see if an existing Blobstore for the
+ * the given BlobStoreKey exists, and if so, use the existing one.
+ * Otherwise a new one will be created and added to the Repository
+ * </p>
+ */
+public class BlobStoreRepository {
```

Comment:
Why is this all static? A broker will have a single instance of the ManagedLedgerFactory. Similarly, it can have a single instance of the repository, or the ManagedLedgerFactory should have the instance.

Why do we even need a repository? Currently there's only one offloader configurable at a time for the broker.

Reply:
This class is intended to replace the internal map of "readBlobStores" that was in the `BlobStoreManagedLedgerOffloader` class.  My first observation was that we can use the same BlobStore for both reading and writing, so we didn't need to have two references to the same blobStore.  

I debated on whether we needed a repository or not, and thought that by adding it now, we could theoretically add support for multiple offloads in the future (perhaps in a backup scenario or similar) where you want to offload to S3 and GCS.  

it may come in handy is we ever wanted to support multi-threaded offloads in the future, perhaps one thread per topic or similar?

Reply:
So add back the internal map of "readBlobStores" to the `BlobStoreManagedLedgerOffloader` class?

Reply:
Now I remember why I added the BlobStoreRepository class in the first place. The other issue that was bothering me was the existence of a "back door" constructor for the BlobStoreManagedLedgerOffloader class that had the @VisibleForTesting annotation. 

It was there because the blobStore had to first be created an in unit tests and then passed in to that constructor. This enabled the ability to both have functional blobStore and spy/mock it at the same time.  

I wasn't a big fan of testing this way, as it was a different code path that we were using than what would actually be executed at runtime. Therefore, I refactored it out, and discovered that without a way to retain the blobStore instances between unit tests, that all of the unit tests broke. So the static BlobStoreRepository class was created to fix all of the unit tests that were broke from removing that one constructor. At this point, removing it would break all of the unit tests.

Reply:
What you describe, with passing the blobStore instance in is standard inversion of control or dependency injection. In general, it's a good thing to have, especially for unit testing.

This repository is equivalent to a global variable.

I think this should be changed back to use DI. Your concern about the create path not being tested is addressed by the whole thing being tested in the integration test.

Reply:
`    // build context for jclouds BlobStoreContext, mostly used in test
    @VisibleForTesting
    BlobStoreManagedLedgerOffloader(BlobStore blobStore, String container, OrderedScheduler scheduler,
                                    int maxBlockSize, int readBufferSize) {
        this(blobStore, container, scheduler, maxBlockSize, readBufferSize, Maps.newHashMap());
    }

    BlobStoreManagedLedgerOffloader(BlobStore blobStore, String container, OrderedScheduler scheduler,
                                    int maxBlockSize, int readBufferSize,
                                    Map<String, String> userMetadata) {
        this.offloadDriverName = "aws-s3";
        this.scheduler = scheduler;
        this.readBufferSize = readBufferSize;
        this.writeBucket = container;
        this.writeRegion = null;
        this.writeEndpoint = null;
        this.maxBlockSize = maxBlockSize;
        this.writeBlobStore = blobStore;
        this.writeLocation = null;
        this.userMetadata = userMetadata;
        this.credentials = null;

        readBlobStores.put(
            BlobStoreLocation.of(writeRegion, writeEndpoint),
            blobStore
        );
    }
`

This is the method in question that I replaced, it has a lot more issues that just only be used for testing.  There are a lot of values set to NULL, the driver name is hard-coded, which made testing for Azure REALLY difficult.  And lastly, it was also keeping an internal Map of BlobStores, which is what the BlobStoreRepository does. 

Putting this back in will create more issues than it solves.

Reply:
I agree that this constructor is ugly and should get some refactoring. The driver name should be passed in as the first argument. Write bucket, region, endpoint, should all be passed on, maybe even as a map so you don't have to construct the map in getOffloadDriverMetadata each time. 

There should only be one constructor. writeBlobStore should always be constructed outside of BlobStoreManagedLedgerOffloader and passed in.

The internal map of blobstores is a member field of BlobStoreManagedLedgerOffloader. It's not a global map like BlobStoreRepository.

Reply:
I think passing in a map into the constructor is the best approach, as it insulates from a lot of future changes, and allows us to add non-blob-storage JClouds providers in the future, such as [Filesystem](https://jclouds.apache.org/guides/filesystem/) and [Databases](https://jclouds.apache.org/guides/jdbc/)

That is why I chose to use the TieredStorageConfiguration object to encapsulate all of these properties that get passed in to the constructor.  This object also, then delegates to responsibility of instantiating the BlobStore to the JCloudBlobStoreProvider Enumeration object (which is defined in the BlobStoreBuilder interface that all Enums have to implement).  I feel this is a more IoC approach. 

As for the placement of the data structure that holds the BlobStore instances, I guess it doesn't really matter if it is a member field of the BlobStoreManagedLedgerOffloader or not. I can move it there if you like, but my thought process was that by de-coupling the storage of the BlobStores from the offloader gives us the opportunity to re-use them across multiple instances of offloaders in the future, if we were to have multiple providers or even a pool of offloaders running concurrently, as some of these remote connections might be slow and/or timeout, so it might be best to have multiple instances running.

Reply:
> As for the placement of the data structure that holds the BlobStore instances, I guess it doesn't really matter if it is a member field of the BlobStoreManagedLedgerOffloader or not. I can move it there if you like, but my thought process was that by de-coupling the storage of the BlobStores from the offloader gives us the opportunity to re-use them across multiple instances of offloaders in the future, if we were to have multiple providers or even a pool of offloaders running concurrently, as some of these remote connections might be slow and/or timeout, so it might be best to have multiple instances running.

If we want to share like this, then the container should be passed in as an argument to the offloader. However, I don't think it'll be the case where we have multiple instances of the same offloader, at least not in the near-term, so I would suggest you move it into the scope of the BlobStoreManagedLedgerOffloader object, as blobstores was previously.